### ISSUE TITLE
Stabilize Deferred.getCompleted and related

### ISSUE THREAD

**Issue by cy6erGn0m** [Kotlin/kotlinx.coroutines#2656]
At the moment, `Deferred.getCompleted()` and `Deferred.getCompletionExceptionOrNull()` are experimental and marked with `@ExperimentalCoroutinesApi`.

The functions above are useful for polling and accessing the deferred result from a non-suspend context where it's impossible to wait for the completion. The related function `invokeOnCompletion {}` is usually used with these experimental function to access the computation result after its completion. And of course, the callback handler is a regular non-suspend lambda. So without using experimental functions, the only solution is to launch a coroutine in the handler or trigger some existing coroutine (via channel or something). This is too much boilerplate. Fortunately, we have the experimental function providing a way to access the computation result just in place.

The typical usages I've faced are the following.

Generally using the pair looks like:
```kotlin
deferred.invokeOnCompletion { cause ->
    when (cause) {
        null -> deferred.getCompleted()
        else -> ....
    }
}
```

Handle only if successful:
```kotlin
deferred.invokeOnCompletion { cause ->
    if (cause != null) {
        handle(deferred.getCompleted())
    }
}
```

getCompletedOrNull:
```kotlin
deferred.invokeOnCompletion { cause ->
    handle(when (cause) {
        null -> deferred.getCompleted()
        else -> null
    })
}
```

Ensure resource cleanup:
```kotlin
deferred.invokeOnCompletion { cause ->
    if (cause != null) {
        deferred.getCompleted().close()
    }
}
```

Bridge two deferred:
```kotlin
deferred.invokeOnCompletion { cause ->
    when (cause) {
        null -> publicDeferred.complete(deferred.getCompleted())
        else -> publicDeferred.completeExceptionally(cause)
    }
}
```

Tracking instant async job status (e.g. in web UI without websockets and instant notification):
```kotlin
onPageLoad {
    when {
        !deferred.isCompleted -> putInProgressLabel()
        deferred.isCancelled -> putCancelledLabel()
        deferred.getCompletionException() != null -> putFailedLabel()
        else -> putSuccessLabel(deferred.getCompled())
}
```

**Comment by qwwdfsad** [comment]
The draft resolution is the following: leave these functions as experimental, and later[ soft-deprecate](https://youtrack.jetbrains.com/issue/KT-54106) them.

As a replacement, we'll add a member function `getResultOrThrow(): Result<T>` (name TBD and is subject to change) with (potentially) default interface implementation, but it currently is blocked by the corresponding Kotlin issue

**Comment by flaviencharlon-toast** [comment]
What is the plan regarding this? `Deferred.getCompleted()` is still experimental and I don't see any `getResultOrThrow()` or equivalent.

**Comment by NTDV** [comment]
Maybe just remove `@ExperimentalCoroutinesApi`? After all, if `Deferred.getCompleted()` and others have been around for 4 years, that means a lot. Can the planned functionality exist without removing this? Or are there other non-obvious reasons why these functions are destined for removal?
